gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java

```diff
@@ -294,49 +328,145 @@ public void run() {
                                 UInt32Value.newBuilder().setValue(objectHasher.hash().asInt())));
                   }
                 }
-                requestObserver.onNext(requestBuilder.build());
-
-                if (objectFinalized) {
-                  // Close the request half of the streaming RPC.
-                  requestObserver.onCompleted();
+                if (dataChunkMap.size() == numberOfRequestsToRetain) {
+                  dataChunkMap.remove(dataChunkMap.firstKey());
+                  retryMap.remove(dataChunkMap.firstKey());
```

Comment:
Why do we need two different maps from offset to chunk?

Reply:
Second is keep track of how many tries have been made to each offset position. Otherwise the logic could end up trying a failed offset with associated data chunk endlessly.

Reply:
This approach will potentially make us retry too many times (in the worst case we'll retry UPLOAD_RETRIES per chunk) and that the number of retries varies with object size being uploaded.

Let's please keep this simpler. Impose the UPLOAD_RETRIES limit at the top level, i.e., make that the maximum number of times we retry for a transient error. That will be easier to understand and consistent with the way retry handling is implemented in all other GCS clients. And it will let you get rid of this second map, making the code easier to read and reason about.

Reply:
As discussed offline, this map can be removed and will enforce the maximum retries of UPLOAD_RETRIES at top lvl.